packages/core/src/external/sftp/client.ts

```diff
@@ -0,0 +1,136 @@
+import SshSftpClient from "ssh2-sftp-client";
+import { Writable } from "stream";
+
+export interface SftpConfig {
+  host: string;
+  port: number;
+  username: string;
+  password: string;
+  privateKey: string;
+}
+
+export interface SftpClientImpl {
+  connect(): Promise<void>;
+  read(remotePath: string): Promise<Buffer>;
+  write(remotePath: string, content: Buffer): Promise<boolean>;
+  list(remotePath: string): Promise<string[]>;
+  exists(remotePath: string): Promise<boolean>;
+}
+
+export class SftpClient implements SftpClientImpl {
+  protected readonly client: SshSftpClient;
+  protected readonly host: string;
+  protected readonly port: number;
+  protected readonly username: string;
+  protected readonly password: string;
+  protected readonly privateKey: string;
+
+  constructor({ host, port, username, password, privateKey }: SftpConfig) {
+    this.client = new SshSftpClient();
+
+    this.host = host;
+    this.port = port;
+    this.username = username;
+    this.password = password;
+    this.privateKey = privateKey;
+  }
+
+  private async executeWithSshListeners<T, F extends (client: SshSftpClient) => Promise<T>>(
+    executionHandler: F
+  ): Promise<T> {
+    let executionCompleted = false;
+    let executionError: Error | undefined;
+    function errorHandler(error?: Error) {
+      if (error != null && error instanceof Error) {
+        executionError = error;
+      } else if (!executionCompleted) {
+        executionError = new Error("Socket was closed before execution completed");
+      }
+    }
+    this.client.on("error", errorHandler);
+    this.client.on("end", errorHandler);
+    this.client.on("close", errorHandler);
```

Comment:
These are the three possible error handlers that will be executed, which may or may not be called with an `Error` object (despite what the documentation says).

https://github.com/theophilusx/ssh2-sftp-client/blob/0a4540d7132a988c1f9478ce656a33d99257e218/src/index.js#L1450